Implement compound assignment operators for BitSet

WaDelma opened the pull request:
Resolves #24

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/slide-rs/hibitset/28)
<!-- Reviewable:end -->

WaDelma committed:
Added method for adding BitSetLike to BitSet.

WaDelma committed:
Construct iterator with method.

torkleyy reviewed (commented):
I like it! I think it's correct, but I want to take a closer look at it. Don't you like my idea (using compound assignment operators)?

WaDelma commented:
I actually didn't understand what you meant by it and I already had this idea in my head.

torkleyy commented:
Just using |= for the operation.

WaDelma commented:
Ah yeah. I can add the trait implementation.

WaDelma commented:
Hmm.. Should I just implement `&=`, `^=` and `!=` too then?

torkleyy commented:
That's what I would expect before closing the issue. If you only want to implement `|=`, that's fine, too.

WaDelma committed:
Implemented BitOrAssign and BitAndAssign. Added clear method to BitIter.

WaDelma commented:
I implemented `|=` (which was basically just moving `add_all`) and `&=` which was bit more involved.

To implement `&=` I added `clear` method for `BitIter`. Should we use that for the `BitSet::clear`? As far I can see it should perform better, right? Have to benchmark that.

There is still some problems with the PR. I had to duplicate `handle_level` method. I'll try to clean it up.

WaDelma commented:
Also `&=` has efficiency problem: It always grows the rhs of it. I'll have to think how to fix that.

WaDelma committed:
Descend only to children that are on both sets.

WaDelma committed:
Implemented xor assignment.

WaDelma commented:
I think this might be ready for review.

torkleyy reviewed (commented):
I like the code, it's just that I didn't review the logic yet. I'm not sure when I will be able to do so.

WaDelma commented:
So is there anything to improve here?
(Mostly just pinging this so it won't be forgotten)

torkleyy reviewed (approved):
Thanks! Sorry, I forgot this one.

torkleyy commented:
r? @kvark @Aceeri

kvark commented:
Would be nice to have better commit messages

kvark commented on src/iter/mod.rs:
```diff
@@ -38,7 +38,29 @@ impl<T: BitSetLike> BitIter<T> {
+    /// Clears the rest of the bitset that hasn't been iterated yet.
+    pub fn clear(&mut self) {
+        use self::State::*;
+        'find: loop {
```
could this be rewritten as
```rust
while let Some(level) = (1..LAYERS)..find(|&level| self.handle_level(level) == Continue) {
    let lower = level - 1;
    let idx = (self.prefix[lower] >> BITS) as usize;
    *self.set.layer_mut(lower, idx) = 0;
}
```

torkleyy commented:
Reviewed 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/slide-rs/hibitset/28)*
<!-- Sent from Reviewable.io -->

WaDelma committed:
Refactored ops and added tests.

kvark commented on src/ops.rs:
```diff
@@ -10,23 +10,12 @@ impl<'a, B> BitOrAssign<&'a B> for BitSet
-                }
-            }
-            break;
+        while let Some(level) = (1..LAYERS).find(|&level| iter.handle_level(level) == Continue) {
```
wow, glad to see this working. Looks better!

kvark commented on src/iter/mod.rs:
```diff
@@ -111,3 +105,30 @@ impl<T: BitSetLike> BitIter<T> {
+            assert_eq!(0, i);
+        }
+    }
+}
```
nit: newline missing

kvark reviewed (approved):
last commit is better squashed, otherwise good to go

torkleyy commented:
bors delegate+

WaDelma commented:
bors r=torkleyy
